Add udc docs (#526)
OpenZeppelin/cairo-contracts: docs/modules/ROOT/pages/udc.adoc

Account contracts make up very critical components of the StarkNet ecosystem, since a bug in any implementation—let alone a widespread one—could prove disastrous.
Therefore, maximum caution is in order.
In this spirit, we should consider trimming account responsibilities to simplify implementations and minimize bug/attack surfaces.

This standard moves contract deployment responsibilities to an external, specialized deployer contract.
The separation allows accounts to deploy contracts without compromising security.
To remove the redundancy of deploying a new deployer contract for each account, the deployer contract should be a singleton Universal Deployer Contract (UDC).

TIP: For further information on design decisions, see the https://community.starknet.io/t/universal-deployer-contract-proposal/1864[Universal Deployer Contract Proposal].

== Table of Contents
* <<quickstart, Quickstart>>
* <<deploying_the_udc,Deploying the UDC>>
* <<usage,Usage>>
 ** <<reserved_address_space_for_deployer,Reserved Address Space for Deployer>>
 ** <<reproducible_deployments,Reproducible Deployments>>
* <<calculating_counterfactual_addresses,Calculating Counterfactual Addresses>>

> * <<universal_deployer_contract_api,Universal Deployer Contract API>>

Reviewer: I would use `UDC API Specification` for consistency with other parts of the doc. A refactor to unify this pattern (and others) would be helpful IMO (out of scope for this PR).

== Quickstart

Account contracts need to first declare the contract and then send a `deployContract` transaction to the Universal Deployer Contract address.
Here's an example implementation in Python:

[,python]
----
signer = MockSigner(PRIVATE_KEY)

async def deploy_contract():
    # Declare the contract and get the class hash before deployment
    class_hash, _ = await signer.declare_class(account, "Ownable")

    # Set up deployment parameters
    salt = 123456  # random number
    unique = FALSE
    calldata = [account.contract_address]
    params = [class_hash, salt, unique, len(calldata), *calldata]

    # Send a "deployContract" transaction to the deployer contract
    await account.send_transaction(
        account,

>         deployer_address,

Author: Ah right. It's available from Nile. I'll add it in

        "deployContract",
        params
    )
----

> TIP: The UDC address is `0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf`.

Reviewer: ```suggestion
TIP: The UDC address is `0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf` in mainnet, testnets, and starknet-devnet.
```

Note that this address may change in the future.

=== Deploying the UDC

Though, the Universal Deployer Contract has already been deployed, this documentation does not take its deployment for granted.

> The deployment of the UDC requires a contract (most likely an account) to call https://www.cairo-lang.org/docs/hello_starknet/deploying_from_contracts.html#the-deploy-system-call[StarkNet's deploy syscall].

Reviewer: I would add at least a warning, this feature is deprecated and will be removed soon.

The whole point of the UDC lies in eliminating deployment responsibilities from accounts; however, we need to deploy the UDC before relinquishing this functionality from accounts.

The UDC deployment should also set `deploy_from_zero=TRUE` in the `deploy` syscall.
In so doing, this creates a deterministic address that can be used across all instances of StarkNet.

== Usage

> The Universal Deployer Contract offers two deployment types: reserving the address space for the deployer and reproducible deployments.

Author: Good call, that's a much clearer distinction

The `unique` boolean parameter ultimately determines the type of deployment.

CAUTION: When deploying a contract that uses `get_caller_address` in the constructor calldata, remember that the UDC (and not the account) deploys that contract.
Therefore, querying `get_caller_address` in a contract's constructor will return the UDC's address and *not* the account's address.

> === Reserved Address Space for Deployer

Reviewer 2: ```suggestion
=== Reserved address space for deployer
```

> Reserving the address space for the deployer deploys a contract unique to the deployer.
> This prevents squatting on addresses.

Reviewer 2: ```suggestion
Making deployments dependent of the deployer address, users can "reserve" a whole address space from squatting by someone else. Only that deployer can deploy to those addresses.
```

Achieving this type of deployment necessitates that the deployer sets `unique` to `TRUE` in the `deployContract` call.
Under the hood, the function call leverages the deployer's address and creates a hashchain by hashing the deployer's address with the given salt.

To deploy a unique contract address:

[,python]
----
signer = MockSigner(PRIVATE_KEY)

async def deploy_contract():
    # Declare the contract and get the class hash before deployment
    class_hash, _ = await signer.declare_class(account, "Ownable")

    # Set up deployment parameters
    salt = 123456  # random number
    unique = TRUE
    calldata = [account.contract_address]
    params = [class_hash, salt, unique, len(calldata), *calldata]

    # Send a "deployContract" transaction to the deployer contract
    await account.send_transaction(
        account,
        deployer_address,
        "deployContract",
        params
    )
----

> === Reproducible Deployments

Reviewer 2: ```suggestion
=== Reproducible deployments
```

Reproducible deployments create contract addresses independent of the deployer and the UDC instance.

> Instead, only the class hash, salt, and constructor arguments determine the address.

Reviewer 2: ```suggestion
Instead only the class hash, salt, and constructor arguments determine the address.
```

This type of deployment enables redeployments of accounts and known systems across multiple networks.
To deploy a reproducible deployment, set `unique` to `FALSE`.

[,python]
----
signer = MockSigner(PRIVATE_KEY)

async def deploy_contract():
    # Declare the contract and get the class hash before deployment
    class_hash, _ = await signer.declare_class(account, "Ownable")

    # Set up deployment parameters
    salt = 123456  # random number
    unique = FALSE
    calldata = [account.contract_address]
    params = [class_hash, salt, unique, len(calldata), *calldata]

    # Send a "deployContract" transaction to the deployer contract
    await account.send_transaction(
        account,
        deployer_address,
        "deployContract",
        params
    )
----

> == Calculating Counterfactual Addresses

Reviewer 2: ```suggestion
== Calculating counterfactual addresses
```

Reviewer 2: why is this section relevant to the UDC? i think i prefer to keep it because is loosely related to deployments and might come handy, but otherwise i see no relation with the UDC

Author: I was thinking that if a user wanted info on calculating addresses when deploying through the UDC, this would be the logical place to look. I'm not sure if there's a better home for it ATM.

I do agree it's not directly relevant to the UDC. Your call if you'd like to keep it here

= Universal Deployer Contract

> The Universal Deployer Contract (UDC) is a singleton smart contract that wraps the https://www.cairo-lang.org/docs/hello_starknet/deploying_from_contracts.html#the-deploy-system-call[`deploy`] syscall.
> This contract intends to serve and support all other contracts that do not implement this functionality (like most accounts).
> And since accounts lack a native deployment transaction type, the UDC aims to stand in this while also providing a standardized event to ease observability.

Reviewer 2: ```suggestion
The Universal Deployer Contract (UDC) is a singleton smart contract that wraps the https://www.cairo-lang.org/docs/hello_starknet/deploying_from_contracts.html#the-deploy-system-call[`deploy syscall`] to expose it to any contract that doesn't implement it, such as account contracts. You can think of it as a **standardized generic factory for StarkNet contracts**.

And since StarkNet has no deployment transaction type, it offers a canonical way to deploy smart contracts by following the [standard deployer interface](https://community.starknet.io/t/snip-deployer-contract-interface/2772) and emitting a `ContractDeployed` event.
``````
how does this read?

Author: Much better!

TIP: For further information on design decisions, see the https://community.starknet.io/t/universal-deployer-contract-proposal/1864[Universal Deployer Contract Proposal].

> NOTE: The UDC address is `0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf` in mainnet, testnets, and starknet-devnet.
> Note that this address may change in the future.

Reviewer 2: ```suggestion
NOTE: The UDC address is `0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf` in mainnet, testnets, and starknet-devnet. This address may change in the future.
```

it is already a note :p

    # Send a "deployContract" transaction to the deployer contract
    await account.send_transaction(
        account,
        UNIVERSAL_DEPLOYER_ADDRESS,
        "deployContract",
        params
    )
----

=== Deploying the UDC

NOTE: The Universal Deployer Contract has already been deployed on most networks and development environments.
The deployment of the UDC requires a contract (most likely an account) to call https://www.cairo-lang.org/docs/hello_starknet/deploying_from_contracts.html#the-deploy-system-call[StarkNet's deploy syscall].

The UDC deployment should also set `deploy_from_zero=TRUE` in the `deploy` syscall.
In so doing, this creates a deterministic address that can be used across all instances of StarkNet.

== Usage

> The Universal Deployer Contract offers two deployment types: deployer dependent and deployer independent.

Reviewer 2: ```suggestion
The Universal Deployer Contract offers two types of address where to deploy: deployer dependent and deployer independent.
```

> The Universal Deployer Contract offers two deployment types: deployer dependent and deployer independent.

Reviewer 2: ```suggestion
The Universal Deployer Contract offers two types of address where to deploy: deployer dependent and deployer independent.
```

> The Universal Deployer Contract offers two deployment types: deployer dependent and deployer independent.

Reviewer 2: ```suggestion
The Universal Deployer Contract offers two types of address where to deploy: deployer dependent and deployer independent.
```

=== Deploying the UDC

NOTE: The Universal Deployer Contract has already been deployed on most networks and development environments.
The deployment of the UDC requires a contract (most likely an account) to call StarkNet's `deploy` syscall.

The UDC deployment should also set `deploy_from_zero=TRUE` in the `deploy` syscall.

> In so doing, this creates a deterministic address that can be used across all instances of StarkNet.

Reviewer 2: ```suggestion
The deployment of the UDC requires `deploy_from_zero=TRUE` to be passed to the `deploy` syscall. This results in a deterministic and predictable address across all instances of StarkNet, facilitating SDK integration and reproducibility of deployments.
```
